src/cyclefold/gadgets.rs

```diff
@@ -0,0 +1,614 @@
+      n_limbs,
+    )?;
+
+    let u = AllocatedR1CSInstance::alloc(cs.namespace(|| "U"), inst.map(|x| &x.u))?;
+
+    let T = AllocatedPoint::alloc(cs.namespace(|| "T"), inst.map(|x| x.T.to_coordinates()))?;
+    T.check_on_curve(cs.namespace(|| "T on curve"))?;
+
+    Ok(Self { U, u, T })
+  }
+
+  // TODO: Delete if not needed, or uncomment if this saves time
+  // pub fn absorb_in_ro<CS: ConstraintSystem<<E as Engine>::Base>>(
+  //   &self,
+  //   mut cs: CS,
+  //   ro: &mut E::ROCircuit,
+  // ) -> Result<(), SynthesisError> {
+  //   self.U.absorb_in_ro(cs.namespace(|| "absorb U"), ro)?;
+  //   self.u.absorb_in_ro(ro);
+  //   ro.absorb(&self.T.x);
+  //   ro.absorb(&self.T.y);
+  //   ro.absorb(&self.T.is_infinity);
+  //   Ok(())
+  // }
+}
+
+pub mod emulated {
+  use bellpepper::gadgets::Assignment;
+  use bellpepper_core::{
+    boolean::{AllocatedBit, Boolean},
+    num::AllocatedNum,
+  };
+
+  use super::*;
+
+  use std::marker::PhantomData;
+
+  use crate::{
+    constants::{NUM_CHALLENGE_BITS, NUM_FE_FOR_RO},
+    gadgets::{
+      nonnative::{bignat::BigNat, util::f_to_nat},
+      utils::{
+        alloc_bignat_constant, alloc_one, alloc_zero, conditionally_select,
+        conditionally_select_bignat, le_bits_to_num, /*scalar_as_base,*/
+      },
+    },
+    traits::{Group, ROConstantsCircuit},
+    RelaxedR1CSInstance,
+  };
+
+  use ff::Field;
+
+  #[derive(Clone)]
+  pub struct AllocatedPoint<E1, E2>
+  where
+    E1: Engine<Base = <E2 as Engine>::Scalar>,
+    E2: Engine<Base = <E1 as Engine>::Scalar>,
+  {
+    x: BigNat<E1::Base>,
+    y: BigNat<E1::Base>,
+    is_infinity: AllocatedNum<E1::Base>,
+    _p: PhantomData<E2>,
+  }
+
+  impl<E1, E2> AllocatedPoint<E1, E2>
+  where
+    E1: Engine<Base = <E2 as Engine>::Scalar>,
+    E2: Engine<Base = <E1 as Engine>::Scalar>,
+  {
+    pub fn alloc<CS>(
+      mut cs: CS,
+      coords: Option<(E2::Base, E2::Base, bool)>,
+      limb_width: usize,
+      n_limbs: usize,
+    ) -> Result<Self, SynthesisError>
+    where
+      CS: ConstraintSystem<<E1 as Engine>::Base>,
+    {
+      let x = BigNat::alloc_from_nat(
+        cs.namespace(|| "x"),
+        || {
+          Ok(f_to_nat(
+            &coords.ok_or(SynthesisError::AssignmentMissing)?.0,
+          ))
+        },
+        limb_width,
+        n_limbs,
+      )?;
+
+      let y = BigNat::alloc_from_nat(
+        cs.namespace(|| "y"),
+        || {
+          Ok(f_to_nat(
+            &coords.ok_or(SynthesisError::AssignmentMissing)?.1,
+          ))
+        },
+        limb_width,
+        n_limbs,
+      )?;
+
+      let is_infinity = AllocatedNum::alloc(cs.namespace(|| "is_infinity"), || {
+        Ok(if coords.map_or(true, |c| c.2) {
+          E1::Base::ONE
+        } else {
+          E1::Base::ZERO
+        })
+      })?;
+      cs.enforce(
+        || "is_infinity is bit",
+        |lc| lc + is_infinity.get_variable(),
+        |lc| lc + CS::one() - is_infinity.get_variable(),
+        |lc| lc,
+      );
```

Comment:
Why do you not just allocate an `AllocatedBit` instead of manually simulating one?